QUIC Server API: Updates to polling design document

hlandau opened the pull request:
Work on the https://github.com/openssl/project/issues/418 Polling Lite implementation led to some more detailed thought on how event types need to be defined.

Sashan commented on doc/designs/quic-design/server/quic-polling.md:
```diff
@@ -228,40 +228,535 @@ typedef struct bio_poll_descriptor_st {
+  failure of the `SSL_poll()` mechanism itself (`SSL_poll()` returning 0).
+
+### Definitions
+
```
this feels like repeating the same list we started with at line 233.

Sashan reviewed (changes requested):
there are few typos which might be worth to fixing.

I'd like to ask you to consider to reorder things in a document. The order I'm suggesting is:
- start with General Principles
- definitions follow
- discussion with its details

hlandau commented:
Updated

Sashan reviewed (approved):
thank you for moving sections around. it's much easier to read coming to this stuff for the first time.

mattcaswell commented on doc/designs/quic-design/server/quic-polling.md:
```diff
@@ -228,42 +228,507 @@ typedef struct bio_poll_descriptor_st {
+**Q. Should applications be able to listen on `R` but not `ER`?**
+
+A. This would enable an application to listen for more application data but not
+care about stream resets. This can be permitted for now even if it raises osme
```
typo: some

t8m commented on doc/designs/quic-design/server/quic-polling.md:
```diff
@@ -228,42 +228,507 @@ typedef struct bio_poll_descriptor_st {
-#define OSSL_POLL_EVENT_OSU        (1U << 7)
+/* Read/write and associated exception event types. */
+#define SSL_POLL_EVENT_RE           (SSL_POLL_EVENT_R  | SSL_POLL_EVENT_ER)
+#define SSL_POLL_EVENT_WE           (SSL_POLL_EVENT_R  | SSL_POLL_EVENT_ER)
```
Copy&paste: `SSL_POLL_EVENT_R| SSL_POLL_EVENT_ER -> SSL_POLL_EVENT_W | SSL_POLL_EVENT_EW`

kroeckx commented:
It's not really clear to me how we're going to support applications that want to implement their own polling. Does that mean that for linux we'd have to add support for something like eventfd? I guess for Windows we'd have to add IOCP support, and so on?

kroeckx commented:
I guess we currently have no proposal for that, while I think that's more important to have them providing our own kqueue-like API.

hlandau commented:
> It's not really clear to me how we're going to support applications that want to implement their own polling. Does that mean that for linux we'd have to add support for something like eventfd and timerfd? I guess for Windows we'd have to add IOCP support, and so on?

If an application wants to do its own polling, it can do so. It would do so by determining the underlying network resources using `SSL_get_[rw]poll_descriptor` and registering them in its own reactor as per `SSL_net_(read|write)_desired` and `SSL_get_event_timeout`. After it wakes up, it can use either poller API with a timeout of zero in order to extract information on what objects are ready (search 'hierarchical polling' in this document).

hlandau commented:
Updated.

kroeckx commented:
> If an application wants to do its own polling, it can do so. It would do so by determining the underlying network resources using `SSL_get_[rw]poll_descriptor` and registering them in its own reactor as per `SSL_net_(read|write)_desired` and `SSL_get_event_timeout`. After it wakes up, it can use either poller API with a timeout of zero in order to extract information on what objects are ready (search 'hierarchical polling' in this document).

From the discussion in #23339, it seems to me that it's not possible to write a performant implementation with that, and that we're missing functionality to be able to do that.

hlandau commented:
> > If an application wants to do its own polling, it can do so. It would do so by determining the underlying network resources using `SSL_get_[rw]poll_descriptor` and registering them in its own reactor as per `SSL_net_(read|write)_desired` and `SSL_get_event_timeout`. After it wakes up, it can use either poller API with a timeout of zero in order to extract information on what objects are ready (search 'hierarchical polling' in this document).
> 
> From the discussion in #23339, it seems to me that it's not possible to write a performant implementation with that, and that we're missing functionality to be able to do that.

No? I don't believe that discussion has fully ingested the proposal here.

There are two different issues here:

1. Determining when the QUIC engine needs to have internal event processing done. This is addressed by the above.
2. Determining which of an arbitrarily large number of QUIC SSL objects are ready to do useful things after that.

But if I'm missing something here let me know. This document is quite thorough and discusses *many* of the relevant issues, so it is worth reading in full.

kroeckx commented:
So I would imagine that an application would like to write it's main loop so that it calls a single function, and then gets a work item. A work item would be something like:
- There is data available on a socket that can be read
- A timer has expired
- The state of the QUIC SSL object has changed and you can now read from it
- The state of the QUIC SSL object has changed and you can now write to it
- ...

The "sketch B" API would probably provide support for that, but I currently don't see how we provide support for not using that API other than the timer. (The data on the socket is provided by the OS.)

hlandau commented:
> So I would imagine that an application would like to write it's main loop so that it calls a single function, and then gets a work item. A work item would be something like:
> 
>     * There is data available on a socket that can be read
> 
>     * A timer has expired
> 
>     * The state of the QUIC SSL object has changed and you can now read from it
> 
>     * The state of the QUIC SSL object has changed and you can now write to it
> 
>     * ...
> 
> 
> The "sketch B" API would probably provide support for that, but I currently don't see how we provide support for doing that not using the "Sketch B" API other than the timer. (The data on the socket is provided by the OS.)

Essentially:

```c
for (;;) {
    rdesc = SSL_get_rpoll_descriptor(conn);
    wdesc = SSL_get_wpoll_descriptor(conn);
    timeout = SSL_get_event_timeout(conn);

    pollset = {0};
    if (SSL_net_read_desired(conn))
      add rdesc to pollset;
    if (SSL_net_write_desired(conn))
      add wdesc to pollset;

    poll(pollset, timeout);
    SSL_handle_events(conn);

    /* Find out what is now ready using non-blocking poll operation (hierarchical polling) */
    SSL_POLL_ITEM items[...] = { conn, streamA, streamB };
    SSL_poll(items, SSL_POLL_FLAG_NO_HANDLE_EVENTS, timeout=0);
    /* examine items[...].revents to determine ready streams */
}
```

kroeckx commented:
That seems suboptimal to me. I would expect some event mechanism, so just the correct stream(s) get activated, without having to go over all streams. That is, if data is available on the socket, a call should be made that has as effect that the data from the socket is read and processed, and maybe some event is set. It should then go back to fetch the next work item, which might be the event we just set, that could then be read from the stream.

hlandau commented:
> That seems suboptimal to me.

Which is why the registered-mode API is proposed in this document. `SSL_poll` is just for simple and low-radix use cases.

>I would expect some event mechanism, so just the correct stream(s) get activated, without having to go over all streams. That is, if data is available on the socket, a call should be made that has as effect that the data from the socket is read and processed, and maybe some event is set. It should then go back to fetch the next work item, which might be the event we just set, that could then be read from the stream.

Not entirely sure what you're proposing here. Parts of it sound like a callback mechanism, and others sound like some kind of notification queue that an application can pop from.

kroeckx commented:
It's all about using the same queue as epoll already gives you.

hlandau commented:
> It's all about using the same queue as epoll already gives you.

The registered-mode API will obviously have an epoll implementation on the backend for Linux.

kroeckx commented:
I guess a callback mechanism is what I'm looking for, and then let the application deal with it how they want.

t8m commented:
I am not sure a callback mechanism is better than the registered polling API that is planned in the design.

kroeckx commented:
I think a callback will make it a lot easier for applications to adopt. Having to switch to an openssl API will most likely cause various problems, including performance.

hlandau commented:
> I think a callback will make it a lot easier for applications to adopt.

Please elaborate.

It's an explicit objective of our QUIC API design efforts to minimise the changes needed for existing applications to adopt QUIC. Since the existing libssl API design does not use callbacks for I/O event handling (things like certificate verification callbacks are clearly something qualitatively quite different), it doesn't make sense that we would adopt a completely different approach based around I/O callbacks. That would mean existing libssl API users would have to make extremely invasive changes to their code, or completely rewrite it.

Callbacks moreover raise a lot of issues. They don't obviate the need for event handling to be performed when there is work to be done and therefore don't obviate the need for efficient waiting. Any callback-based asynchronous I/O system on *nix is just built on something like epoll anyway, so it doesn't make sense to claim it can be more efficient...

We currently support thread assisted mode in our client API. This means a client can do what it likes without having to worry about a connections's background processing being performed in a timely manner. This means that any callback API would also raise questions regarding a) what thread such an application callback is invoked on (e.g. if thread assisted mode is in use), b) when such a callback is invoked, and whether that invocation will occur in a timely fashion (e.g. if thread assisted mode is not in use and the application is not currently calling libssl), and c) potential reentrancy concerns if an application starts calling libssl APIs within a callback. Of course if an application blocks in such a callback everything falls apart also, but I think that's well understood.

Speaking more generally I don't like callback-intensive API designs because they take control away from the application, and because the application isn't in full control of when they occur and they could occur at "any" time. Any processing an application then does in that callback is also inevitably subject to a variety of caveats and restrictions. It also can't accommodate simpler applications that want to use a blocking-style API. For this reason I tend to avoid callback-driven APIs (in which call arrows go from lower layer to upper layer) and prefer explicit state machines (in which call arrows always go from upper layer to lower layer). These have the advantage of explicit and deterministic processing and synchronisation, explicit and deterministic transfer of control, and explicit and deterministic transfer of any asynchronous notifications that may have occurred to the application.

But that is just my personal design philosophy. It doesn't rule out us offering some kind of callback-style functionality if there are people explicitly saying it would make things easier for their own application architecture (see the curl discussion), *and* the design issues raised above can be resolved satisfactorily. But I absolutely see that as an optional alternative API, not the primary or initial offering.

>Having to switch to an openssl API will most likely cause various problems, including performance.

Please elaborate. Where do you see the performance issues here?

kroeckx commented:
I have a new idea that might work. (Maybe that was yours too, if so that wasn't clear.)

I suggest we have a queue of events per quic connection. So after calling a function that did something, you just check the queue for events, but only for that connection.

We probably want a new function for processing incoming data from the socket. SSL_read is not what you want most of the time, there might not be any user data, and quic probably makes it even more confusing with acks and streams.

Possibly we might want to have our queue of events also directly return a pointer and size of the data that's available.

hlandau commented:
>I suggest we have a queue of events per quic connection. So after calling a function that did something, you just check the queue for events, but only for that connection. We probably want a new function for processing incoming data from the socket. SSL_read is not what you want most of the time, there might not be any user data, and quic probably makes it even more confusing with acks and streams.
>
>Possibly we might want to have our queue of events also directly return a pointer and size of the data that's available.

Yeah. The idea of some kind of "notification queue" is something I've considered and we may end up with something like that at some point. A significant piece of prior art here is the extensions to the Berkeley sockets API for SCTP &mdash; see https://www.rfc-editor.org/rfc/rfc6458, section 6.

Even if we stick to a poll()/read()-like approach for consuming streams, the reality is there's probably going to be "interesting" events on a QUIC connection which an application might need to be notified of (e.g. connection migration). So we will probably end up with some kind of notification mechanism.

Whether notifications are issued to an application by a callback, or by an application calling some "pop notification" function, is a fairly minor detail (we could offer both).

hlandau commented:
@kroeckx This is a valuable discussion so I've moved it to a separate issue: https://github.com/openssl/openssl/issues/24125

I think this PR is ready. Rebased and updated.

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Merged to the feature branch. Thank you.
